src/object_store/src/object/opendal_engine/opendal_object_store.rs

```diff
@@ -0,0 +1,386 @@
+        let key = path.to_string();
+        let last_modified = match opendal_metadata.last_modified() {
+            Some(t) => t.unix_timestamp() as f64,
+            None => 0_f64,
+        };
+
+        let total_size = opendal_metadata.content_length() as usize;
+        let metadata = ObjectMetadata {
+            key,
+            last_modified,
+            total_size,
+        };
+        Ok(metadata)
+    }
+
+    async fn delete(&self, path: &str) -> ObjectResult<()> {
+        self.op.object(path).delete().await?;
+        Ok(())
+    }
+
+    /// Deletes the objects with the given paths permanently from the storage. If an object
+    /// specified in the request is not found, it will be considered as successfully deleted.
+    async fn delete_objects(&self, paths: &[String]) -> ObjectResult<()> {
+        for path in paths {
+            self.op.object(path).delete().await?;
+        }
+        Ok(())
+    }
+
+    async fn list(&self, prefix: &str) -> ObjectResult<Vec<ObjectMetadata>> {
+        let mut object_lister = self.op.object(prefix).list().await?;
+        let mut metadata_list = vec![];
+        while let Some(obj) = object_lister.next().await {
+            let object = obj?;
+            let key = prefix.to_string();
+            let last_modified = match object.metadata().await?.last_modified() {
+                Some(t) => t.unix_timestamp() as f64,
+                None => 0_f64,
+            };
+
+            let total_size = object.metadata().await?.content_length() as usize;
+            let metadata = ObjectMetadata {
+                key,
+                last_modified,
+                total_size,
+            };
+            metadata_list.push(metadata);
+        }
+        Ok(metadata_list)
+    }
+
+    fn store_media_type(&self) -> &'static str {
+        match self.engine_type {
+            EngineType::Memory => "Memory",
+            EngineType::Hdfs => "Hdfs",
+        }
+    }
+}
+
+/// Store multiple parts in a map, and concatenate them on finish.
+pub struct HdfsStreamingUploader {
+    op: Operator,
+    path: String,
+    buffer: FuturesCursor<Vec<u8>>,
+    length: u64,
+}
+impl HdfsStreamingUploader {
+    pub fn new(op: Operator, path: String) -> Self {
+        Self {
+            op,
+            path,
+            buffer: FuturesCursor::new(vec![]),
+            length: 0,
+        }
+    }
+}
+#[async_trait::async_trait]
+impl StreamingUploader for HdfsStreamingUploader {
+    async fn write_bytes(&mut self, data: Bytes) -> ObjectResult<()> {
+        self.buffer.seek(SeekFrom::Start(self.length)).await?;
+        let data = data.to_vec();
+
+        self.length += data.len() as u64;
+        let _ = self.buffer.write(&data).await?;
+        Ok(())
+    }
+
+    async fn finish(mut self: Box<Self>) -> ObjectResult<()> {
+        self.buffer.flush().await?;
+        let buffer = self.buffer.get_mut().to_vec();
+        let r = FuturesCursor::new(buffer);
+        self.op
```

Comment:
The current implementation uploads the data on `finish` so it is essientially a "batch" uploader, similar to the regular `upload`. A real streaming uploader should streamline writes on `write_bytes`. In other words, what we need is a "async writer", which can be constructed on `StreamingUploader::new` and used in `write_bytes`. However, after checking the opendal interface and implementation again, it doesn't seem that they have such interface.

Reply:
> However, after checking the opendal interface and implementation again, it doesn't seem that they have such interface.

OpenDAL will implement `ObjectWriter` to make this possible.